File: java/src/main/java/com/example/cote/PGMRS/다리를_지나는_트럭.java
package com.example.cote.PGMRS;


import java.util.*;
import java.nio.file.*;
import java.io.*;

public class 다리를_지나는_트럭 {

	private static final int MAX_BRIDGE_LENTH = 10000;
	private static final int MAX_WEIGHT = 10000;
	private static final int MAX_TRUCK_WEIGHTS_LENGTH = 10000;

	
	private static int solve(int bridgeLength, int weight, int[] truckWeights){

		Queue<Integer> bridge = new LinkedList<>();
		for(int i=0; i<bridgeLength; i++){
			bridge.offer(0);
		}
		int bridgeWeight = 0;
		int count = 0;
		int cur = 0;

		while (cur < truckWeights.length){
			
			bridgeWeight -= bridge.poll();
			if( bridgeWeight + truckWeights[cur]  <= weight){
				bridge.add(truckWeights[cur]);
				bridgeWeight += truckWeights[cur];
				cur++;
			}else {
				bridge.add(0);
			}

			count++;
		}

		while( bridgeWeight > 0){
			bridgeWeight -= bridge.poll();
			count++;
		}


		return count;
	}

	public static void main(String... args) throws IOException{

		Path p = Paths.get(System.getProperty("user.dir")+"/data/다리를_지나는_트럭.txt");
		BufferedReader rd = Files.newBufferedReader(p);

		int testCases = Integer.parseInt(rd.readLine());
		while(testCases > 0){

			int bridgeLength = Integer.parseInt(rd.readLine());
			int weight = Integer.parseInt(rd.readLine());
			int[] truckWeights = Arrays.stream(rd.readLine().split(" "))
										.mapToInt(Integer::parseInt)
										.toArray();
			int answer = Integer.parseInt(rd.readLine());

			System.out.println(solve(bridgeLength, weight, truckWeights)+ " " + answer);
			testCases--;
		}

	}	
}

File: kotlin/src/main/kotlin/com/example/cote/PGMRS/다리를_지나는_트럭.kt
package com.example.cote.PGMRS

import java.io.*
import java.nio.file.*
import java.util.*

object 다리를_지나는_트럭 {

    val MAX_BRIDGE_LENGTH = 10000
    val MAX_WEIGHT = 10000
    val MAX_TRUCT_WIEGHTS_LENGTH = 10000

    fun solve(bridgeLength: Int, weight: Int, truckWeights: Array<Int>): Int {

        val bridge = ArrayDeque(List(bridgeLength) { 0 })
        var bridgeWeight = 0
        var cur = 0
        var count = 0

        while (cur < truckWeights.size) {

            bridgeWeight -= bridge.removeFirst()

            if (bridgeWeight + truckWeights[cur] <= weight) {
                bridge.addLast(truckWeights[cur])
                bridgeWeight += truckWeights[cur]
                cur++
            } else {
                bridge.addLast(0)
            }

            count++
        }

        while (bridgeWeight > 0) {
            bridgeWeight -= bridge.removeFirst()
            count++
        }

        return count
    }

    fun main() {
        val p =
            Paths.get(System.getProperty("user.dir") + "/data/다리를_지나는_트럭.txt")
        val reader = Files.newBufferedReader(p)

        repeat(reader.readLine().toInt()) {
            val bridgeLength = reader.readLine().toInt()
            val weight = reader.readLine().toInt()
            val truckWeights =
                reader.readLine().split(" ").map(String::toInt).toTypedArray()
            val answer = reader.readLine().toInt()

            println("solved: " + solve(bridgeLength, weight, truckWeights))
            println("answer: " + answer)
        }

        reader.close()
    }
}

fun main() {
    다리를_지나는_트럭.main()
}
